cmd/dagger/functions.go

```diff
@@ -0,0 +1,546 @@
+			subCmd := fc.makeSubCmd(ctx, dag, &modFunction{Name: field.Name, Description: field.Description, ReturnType: field.TypeDef}, c)
+			cmd.AddCommand(subCmd)
+		}
+	}
+}
+
+func (fc *FuncCommand) makeSubCmd(ctx context.Context, dag *dagger.Client, fn *modFunction, c *callContext) *cobra.Command {
+	newCmd := &cobra.Command{
+		Use:   cliName(fn.Name),
+		Short: fn.Description,
+		RunE: func(cmd *cobra.Command, args []string) (err error) {
+			cmd.SetContext(ctx)
+			rec := progrock.FromContext(ctx)
+
+			c.mod.LoadObject(fn.ReturnType)
+
+			for _, arg := range fn.Args {
+				c.mod.LoadObject(arg.TypeDef)
+			}
+
+			argValues := make(map[string]any)
+
+			for _, arg := range fn.Args {
+				p, err := arg.AddFlag(cmd.Flags(), dag)
+				if err != nil {
+					return fmt.Errorf("set flag %q: %w", arg.FlagName(), err)
+				}
+				argValues[arg.Name] = p
+
+				// TODO: This won't work on its own because cobra checks required
+				// flags before the RunE function is called. We need to manually
+				// check the flags after parsing them.
+				if !arg.TypeDef.Optional {
+					cmd.MarkFlagRequired(arg.FlagName())
+				}
+			}
+
+			if fc.OnSetFlags != nil {
+				return fc.OnSetFlags(cmd, fn.ReturnType)
+			}
+
+			if err := cmd.ParseFlags(args); err != nil {
+				return fmt.Errorf("parse sub-command flags: %w", err)
+			}
+
+			fc.addSubCommands(ctx, dag, fn.ReturnType.AsObject, c, cmd)
+
+			// Need to make the query selection before chaining off.
+			returnType, err := fc.selectFunc(fn, c, argValues, dag)
+			if err != nil {
+				return fmt.Errorf("query selection: %w", err)
+			}
+
+			help, _ := cmd.Flags().GetBool("help")
+			cmdArgs := cmd.Flags().Args()
+
+			subCmd, subArgs, err := cmd.Find(cmdArgs)
+			if err != nil {
+				return fmt.Errorf("find sub-command: %w", err)
+			}
+
+			// If a subcommand was found, run it instead.
+			if subCmd != cmd {
+				return fc.run(subCmd, subArgs)
+			}
+
+			if fc.CheckReturnType != nil {
+				if err = fc.CheckReturnType(c, returnType); err != nil {
+					return err
+				}
+			}
+
+			c.loader.Complete()
+
+			vtx := rec.Vertex(
+				"cmd-func-exec",
+				fmt.Sprintf("running %q", cmd.CommandPath()),
+				progrock.Focused(),
+			)
+			defer func() { vtx.Done(err) }()
+
+			setCmdOutput(cmd, vtx)
+
+			// Make sure --help is running on the leaf command (subCmd == cmd),
+			// otherwise the first command in the chain swallows it.
+			if help {
+				return cmd.Help()
+			}
+
+			query, _ := c.q.Build(ctx)
+			rec.Debug("executing", progrock.Labelf("query", "%+v", query))
+
+			var response any
+
+			q := c.q.Bind(&response)
+
+			if err := q.Execute(ctx, c.c); err != nil {
+				return err
+			}
+
+			if fc.OnResult != nil {
+				return fc.OnResult(c, cmd, *returnType, &response)
+			}
+
+			if returnType.Kind != dagger.Voidkind {
+				cmd.Println(response)
+			}
+
+			return nil
+		},
+	}
+
+	if fn.Name != newCmd.Name() {
+		newCmd.Aliases = append(newCmd.Aliases, fn.Name)
+	}
```

Comment:
Yes, it adds only one alternative casing per command, if it's different from the kebab case. This only shows when calling `--help` on that command:
```
$ dagger call -m github.com/shykes/daggerverse/helloWorld --silent with-name --help
Change the name

Usage:
  dagger call with-name [flags]
  dagger call with-name [command]

Aliases:
  with-name, WithName

Available Commands:
  message       Say hello to the world!
  with-greeting Change the greeting
  with-name     Change the name
```

I made it an alias to be easier because all the args parsing is delegated to cobra. If not an alias the easiest thing would be to manicure all the args to kebab case before parsing.

Reply:
Okay cool, looking at the output now that seems totally fine. I previously imagined some crazy thing of every permutation of casing for every name, but in practice it looks good and is helpful to the user 👍